File: aoj/16/aoj1603.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long int ll;
#define FOR(i,s,x) for(int i=s;i<(int)(x);i++)
#define REP(i,x) FOR(i,0,x)
#define ALL(c) c.begin(), c.end()
#define DUMP( x ) cerr << #x << " = " << ( x ) << endl

const int dr[4] = {-1, 0, 1, 0};
const int dc[4] = {0, 1, 0, -1};

pair<int, int> dp[101][50001];
bool used[101][50001];
int N, P[100];

pair<int, int> dfs(int idx, int coins) {
  if (idx == N) return make_pair(0, 0);
  if (used[idx][coins]) return dp[idx][coins];

  used[idx][coins] = true;
  // not buy;
  pair<int, int> ret = dfs(idx + 1, coins);

  // buy only with bills
  int change = (1000 - P[idx] % 1000) % 1000;
  int cnt, money; tie(cnt, money) = dfs(idx + 1, coins + change % 500);
  ret = max(ret, make_pair(cnt + (change >= 500), money - P[idx]));

  // buy with bills & coins to get 500
  if ((P[idx] + 500) % 1000 <= coins) {
    int cnt, money; tie(cnt, money) = dfs(idx + 1, coins - (P[idx] + 500) % 1000);
    ret = max(ret, make_pair(cnt + 1, money - P[idx]));
  }

  return dp[idx][coins] = ret;
}

int main() {
  // use scanf in CodeForces!
  cin.tie(0);
  ios_base::sync_with_stdio(false);

  while (true) {
    cin >> N;
    if (not N) break;
    REP(i, N) cin >> P[i];
    memset(used, 0, sizeof(used));
    int cnt, money; tie(cnt, money) = dfs(0, 0);
    cout << cnt << ' ' << -money << endl;
  }

  return 0;
}

File: aoj/16/aoj1603.py
import sys
if sys.version[0] == '2':
    range, input = xrange, raw_input
sys.setrecursionlimit(10 ** 6)
MAX_COINS = 500 * 100
INF = 10 ** 9


def dfs(idx, coins):
    if idx == N:
        return 0, 0
    elif dp[idx][coins] != -1:
        return dp[idx][coins]

    ret = (0, -INF)

    # not buy
    ret = max(ret, dfs(idx + 1, coins))

    # buy using only bills
    change = (1000 - P[idx] % 1000) % 1000
    cand, money = dfs(idx + 1, coins + change % 500)
    ret = max(ret, (cand + (change >= 500), money - P[idx]))

    # buy using both and get 500
    if (P[idx] + 500) % 1000 <= coins:
        cand, money = dfs(idx + 1, coins - (P[idx] + 500) % 1000)
        ret = max(ret, (cand + 1, money - P[idx]))
    dp[idx][coins] = ret
    # print(idx, coins, ret)
    return ret


while True:
    N = int(input())
    if not N:
        break
    P = [int(input()) for _ in range(N)]

    dp = [[-1] * (MAX_COINS + 1) for _ in range(N + 1)]
    ans, money = dfs(0, 0)
    print(ans, -money)
